design/multiple-storage-types-support.md

```diff
@@ -323,88 +297,131 @@ Some potential ideas for additional types to support in Rook:
+* backend specific images: to avoid image bloat in the main `rook/rook` image, each backend will have their own image that contains all of their daemons and tools.  These images will be used for the various daemons/components of each backend, e.g. `rook/ceph`, `rook/minio`, etc.
+
+#### Layout
+
+The current `rook/rook` repository layout appears to be sufficiently factored to enable multiple storage backend support.
+Some additional directories will be added to support new API versions, new custom resource types, and new storage backends.
+A source code layout that includes these new additions is shown below, annotated with comments about the use of each important directory:
+
+```console
+- cmd # binaries with main entry points
+  - rook # main command entry points for operators and daemons
+    - ceph
+    - minio
+  - rookflex
+- pkg
+  - apis
+    - rook.io # rook.io API group, additional groups would be sibling dirs
+      - v1alpha1 # existing version of alpha Rook API
+      - v1alpha2 # new version of alpha Rook API that includes new types
+    - cluster.rook.io  # common abstractions of cluster, file, object, pool, etc.
+      - v1alpha1
+    - ceph.cluster.rook.io  # ceph specific specs for cluster, file, object
+      - v1alpha1
+    - minio.cluster.rook.io # minio specific specs for cluster, object
+      - v1alpha1
+  - client
+    - clientset # generated strongly typed client code to access Rook APIs
+  - daemon # daemons for each storage backend
+    - ceph
+    - minio
+  - operator # all orchestration logic
+    - cluster # custom controllers for each storage backend
+      - ceph
+      - minio
+```
+
 ## Examples
 
 This section contains concrete examples of storage clusters as a user would define them using yaml.
-In addition to the Ceph and Minio distributed storage cluster examples, we will be considering support for additional storage types in the near future.
+In addition to distributed storage clusters, we will be considering support for additional storage types in the near future.
 
 `ceph-cluster.yaml`:
 
 ```yaml
-apiVersion: rook.io/v1alpha2
-kind: Ceph
+apiVersion: cluster.rook.io/v1alpha2
+kind: StorageCluster
 metadata:
   name: ceph
   namespace: rook-ceph
 spec:
-  monCount: 3
-  storeConfig:
-    storeType: bluestore
-    databaseSizeMB: 1024
-  metadataDevice: nvme01
-  cluster:
-    version: "rook/ceph:v0.1.0"
-    dataDirHostPath: "/var/lib/rook-ceph"
-    hostNetwork: false
-    placement:
-    - name: "mon"
-      nodeAffinity:
-      podAffinity:
-      podAntiAffinity:
-      tolerations:
-    resources:
-    - name: osd
-      limits:
-        cpu: "500m"
-        memory: "1024Mi"
-      requests:
-        cpu: "500m"
-        memory: "1024Mi"
-    scope:
-      deviceFilter: "^sd."
-      location:
+  ceph: # backend specific subsection
```

Comment:
Using the metacontroller, what does the sequence of events look like? Something like this? I feel like I'm missing something, this seems awkward.
1. kubectl create -f ceph-cluster.yaml
2. the rook storage controller gets the event for the cluster creation
3. The rook storage controller creates the ceph crd from the embedded type
4. The ceph controller is notified of the ceph cluster to create
5. The ceph controller creates its cluster.
    - Does the ceph controller already have access to the storage cluster properties? Or does it have to look them up?